drivers: clock_control: add on/off stubs to clock,agilex

fkokosinski opened the pull request:
This PR fixes a regression caused by: 88830a3b.

Due to the stubs not being present in the clock ctrl driver, the ns16550 driver jumped to offset 0x0 when referring to `api->on`.

Fixes #86062.

nordic-krch commented on drivers/clock_control/clock_agilex.c:
```diff
@@ -35,7 +35,19 @@ static int clk_get_rate(const struct device *dev,
+	ARG_UNUSED(sys);
+
+	/* On/off functionality is not supported */
+	return 0;
```
shouldn't be -ENOTSUP?

fkokosinski replied:
That's a tricky question, honestly, and it depends on how we look at trivial API calls implementations.

`-ENOTSUP` is defined as an error code for _unsupported values_: https://docs.zephyrproject.org/apidoc/latest/group__system__errno.html#ga91457bbf35f0f1085619a99423bb1f33, and there’s nothing really wrong with the `dev` or `sys` values that are passed through this call.

The `clock_control_on` call reserves negative return values for errors: https://docs.zephyrproject.org/apidoc/latest/group__clock__control__interface.html#gaec69b0989cefad79ffd5c92f060150b9

> Returns
>	0 on success, negative errno on failure.

Depending on how you look at this, one could argue that a trivial implementation is not really ending in a failure. We can of course add a `-ENOTSUP` here, and modify the NS16550 driver code to not error out upon receiving that errno from the clock control subsystem if that's what you think is better for the overall code organization.

Let me know what you think.

fabiobaltieri replied:
Or maybe the API should return -ENOSYS if it's not implemented and then the caller can handle it in the same way as it handles `EALREADY`?

Anyway sorry folks, this should have been fixed before the release but I'm afraid it won't make the cut since there's still some ambiguity on the api, I've set the milestone to 4.2 and queued it for backport. Still open to discuss it if you can agree of how it should be fixed before EoD.

fkokosinski replied:
@fabiobaltieri `-ENOSYS` sounds good. We've applied necessary changes to the NS16550 driver as well so that it handles `-ENOSYS` the same way it does `-EALREADY`.

fkokosinski replied:
@fabiobaltieri Hard agree here. I've moved returning ENOSYS to the API header.

laudominik committed:
drivers: serial: ns16550: allow clock_control_on returning ENOSYS

Fixes a regression caused by: 88830a3b
The ns16550 driver should handle possibility of
clock_control_on being a stub (i.e. returning ENOSYS)

Signed-off-by: Dominik Lau <dlau@antmicro.com>

fkokosinski committed:
include/zephyr/drivers/clock_control: return `-ENOSYS` for empty `on`/`off`

There are in-tree clock controllers that don't implement the `on`/`off`
calls, e.g. in Intel Agilex. Let's return `-ENOSYS` when that's the case.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
Hey @nordic-krch - could you PTAL at this PR? Thanks!
